incremental_parser/incremental_parser.c

```diff
@@ -0,0 +1,338 @@
+{
+    PARSER_KEYWORD  *tmp_keyword;
+
+    if (strcmp(keyword, "_read") == 0) {
+        working_parser->read_function = (void *) func;
+        return 0;
+    }
+
+    if (strcmp(keyword, "_eof") == 0) {
+        working_parser->eof_function = (void *) func;
+        return 0;
+    }
+
+    if (strcmp(keyword, "_unknown") == 0) {
+        working_parser->unknown_function = (void *) func;
+        return 0;
+    }
+
+    uint32_t    keyword_hash = simple_hash(keyword);
+
+    tmp_keyword = working_parser->keyword;
+
+    while (tmp_keyword) {
+        if (tmp_keyword->keyword_hash == keyword_hash && (!strcmp(tmp_keyword->keyword, keyword))) {
+                if (tmp_keyword->func_no == PARSER_MAX_CALLBACKS)
+                    return 0;
+                tmp_keyword->func[tmp_keyword->func_no++] = (void *) func;
+                return tmp_keyword->func_no;
+        }
+        tmp_keyword = tmp_keyword->next;
+    }
+
+    tmp_keyword = callocz(1, sizeof(*tmp_keyword));
+
+    tmp_keyword->keyword = strdupz(keyword);
+    tmp_keyword->keyword_hash = keyword_hash;
+    tmp_keyword->func[tmp_keyword->func_no++] = (void *) func;
+
+    tmp_keyword->next = working_parser->keyword;
+    working_parser->keyword = tmp_keyword;
+    return tmp_keyword->func_no;
+}
+
+/*
+ * Cleanup a previously allocated parser
+ */
+
+void parser_destroy(INCREMENTAL_PARSER *working_parser)
+{
+    if (unlikely(!working_parser))
+        return;
+
+    PARSER_KEYWORD  *tmp_keyword, *tmp_keyword_next;
+    PARSER_DATA     *tmp_parser_data, *tmp_parser_data_next;
+    
+    // Remove keywords
+    tmp_keyword = working_parser->keyword;
+    while (tmp_keyword) {
+        tmp_keyword_next = tmp_keyword->next;
+        freez(tmp_keyword->keyword);
+        freez(tmp_keyword);
+        tmp_keyword =  tmp_keyword_next;
+    }
+    
+    // Remove pushed data if any
+    tmp_parser_data = working_parser->data;
+    while (tmp_parser_data) {
+        tmp_parser_data_next = tmp_parser_data->next;
+        freez(tmp_parser_data->line);
+        freez(tmp_parser_data);
+        tmp_parser_data =  tmp_parser_data_next;
+    }
+
+    freez(working_parser);
+    return;
+}
+
+
+/*
+ * Fetch the next line to process
+ *
+ */
+
+int parser_next(INCREMENTAL_PARSER *working_parser)
+{
+    char    *tmp = NULL;
+
+    if (unlikely(!working_parser))
+        return 1;
+
+    working_parser->flags &= ~(PARSER_INPUT_PROCESSED);
+
+    PARSER_DATA  *tmp_parser_data = working_parser->data;
+
+    if (unlikely(tmp_parser_data)) {
+        strncpyz(working_parser->buffer, tmp_parser_data->line, PLUGINSD_LINE_MAX);
+        working_parser->data = tmp_parser_data->next;
+        freez(tmp_parser_data->line);
+        freez(tmp_parser_data);
+        return 0;
+    }
+
+    if (unlikely(fileno((FILE *)working_parser->input) == -1)) {
+        error("Stream is no longer valid");
+        return 1;
+    }
```

Comment:
This doesn't seem to allow working with `char *` lines.

Reply:
Removed